MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs

```diff
@@ -940,6 +941,129 @@ private void OnSetAnimationState(SetAnimationState payload)
+                if (options.Pitch != null)
+                {
+                    //convert from halftone offset (-12/0/12/24/36) to pitch multiplier (0.5/1/2/4/8).
+                    soundInstance.pitch = Mathf.Pow(2.0f, (options.Pitch.Value / 12.0f));
+                }
+                if (options.Looping != null)
+                {
+                    soundInstance.loop = options.Looping.Value;
+                }
+            }
+        }
+
+
+        public static Dictionary<Guid, AudioSource> _soundInstances = new Dictionary<Guid, AudioSource>();
+        public static List<Guid> _unpausedSoundInstances = new List<Guid>();
+        int soundStoppedCheckIndex= 0;
+
+        private void SoundUpdate()
+        {
+            //garbage collect expired sounds, one per frame
+            if (soundStoppedCheckIndex >= _unpausedSoundInstances.Count)
+            {
+                soundStoppedCheckIndex = 0;
+            }
+            else
+            {
+                var id = _unpausedSoundInstances[soundStoppedCheckIndex];
+                var soundInstance = _soundInstances[id];
+                if (!soundInstance.isPlaying)
+                {
+                    DestroySoundInstance(soundInstance, id);
+                }
+                else
+                {
+                    soundStoppedCheckIndex++;
+                }
+            }
+        }
+
+        private void DestroySoundInstance(AudioSource soundInstance, Guid id)
+        {
+            Component.Destroy(soundInstance);
+            _soundInstances.Remove(id);
+            var index = _unpausedSoundInstances.FindIndex(x => x == id);
+            if (index >= 0)
+            {
+                _unpausedSoundInstances.RemoveAt(index);
+            }
+        }
+        [CommandHandler(typeof(SetSoundState))]
+        private void OnSetSoundState(SetSoundState payload)
+        {
+            var actor = _actorManager.FindActor(payload.ActorId);
+            if (actor != null)
+            {
+                if (payload.SoundCommand == SoundCommand.Start)
+                {
+                    var obj = MREAPI.AppsAPI.AssetCache.GetAsset(payload.SoundAssetId);
+                    var audioClip = MREAPI.AppsAPI.AssetCache.GetAsset(payload.SoundAssetId) as AudioClip;
+                    if (audioClip != null)
+                    {
+                        var soundInstance = actor.gameObject.AddComponent<AudioSource>();
+                        soundInstance.clip = audioClip;
+                        soundInstance.time = payload.StartTimeOffset;
+                        ApplySoundStateOptions(soundInstance, payload.Options);
+                        soundInstance.Play();
+                        _soundInstances.Add(payload.Id, soundInstance);
+                        _unpausedSoundInstances.Add(payload.Id);
+                    }
+                }
+                else
+                {
+                    var soundInstance = _soundInstances[payload.Id];
```

Comment:
Is there cases where this id coming in the payload could not be in this dictionary?  If so, this should raise a KeyNotFound exception, and should be avoided by using the contains key test around this or the TryGetValue method.

Reply:
fixed everywhere

---
In reply to: [260559537](https://github.com/Microsoft/mixed-reality-extension-unity/pull/35#discussion_r260559537) [](ancestors = 260559537)